File: ScapesCore/src/main/java/org/tobi29/scapes/server/format/basic/BasicWorldFormat.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.server.format.basic;

import java8.util.Optional;
import java8.util.function.Function;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.chunk.EnvironmentServer;
import org.tobi29.scapes.chunk.IDStorage;
import org.tobi29.scapes.chunk.WorldServer;
import org.tobi29.scapes.chunk.terrain.infinite.TerrainInfiniteServer;
import org.tobi29.scapes.engine.utils.io.filesystem.FilePath;
import org.tobi29.scapes.engine.utils.io.filesystem.FileUtil;
import org.tobi29.scapes.engine.utils.io.tag.TagStructure;
import org.tobi29.scapes.plugins.PluginFile;
import org.tobi29.scapes.plugins.Plugins;
import org.tobi29.scapes.server.ScapesServer;
import org.tobi29.scapes.server.format.PlayerData;
import org.tobi29.scapes.server.format.TerrainInfiniteFormat;
import org.tobi29.scapes.server.format.WorldFormat;

import java.io.IOException;
import java.security.AccessController;
import java.security.PrivilegedAction;
import java.security.PrivilegedActionException;
import java.security.PrivilegedExceptionAction;
import java.util.ArrayList;
import java.util.List;

public class BasicWorldFormat implements WorldFormat {
    protected static final Logger LOGGER =
            LoggerFactory.getLogger(BasicWorldFormat.class);
    protected final FilePath path, regionPath;
    protected final IDStorage idStorage;
    protected final Plugins plugins;
    protected final PlayerData playerData;
    protected final TagStructure worldsTagStructure;
    protected final long seed;

    public BasicWorldFormat(FilePath path, TagStructure tagStructure)
            throws IOException {
        SecurityManager security = System.getSecurityManager();
        if (security != null) {
            security.checkPermission(
                    new RuntimePermission("scapes.worldFormat"));
        }
        this.path = path;
        seed = tagStructure.getLong("Seed");
        idStorage = new IDStorage(tagStructure.getStructure("IDs"));
        worldsTagStructure = tagStructure.getStructure("Worlds");
        regionPath = path.resolve("region");
        playerData = new BasicPlayerData(path.resolve("players"));
        plugins = createPlugins();
    }

    protected Plugins createPlugins() throws IOException {
        return new Plugins(pluginFiles(), idStorage);
    }

    @Override
    public IDStorage idStorage() {
        return idStorage;
    }

    @Override
    public PlayerData playerData() {
        return playerData;
    }

    @Override
    public long seed() {
        return seed;
    }

    @Override
    public Plugins plugins() {
        return plugins;
    }

    @SuppressWarnings("CallToNativeMethodWhileLocked")
    @Override
    public synchronized WorldServer registerWorld(ScapesServer server,
            Function<WorldServer, EnvironmentServer> environmentSupplier,
            String name, long seed) throws IOException {
        Optional<TerrainInfiniteFormat> format = AccessController.doPrivileged(
                (PrivilegedAction<Optional<TerrainInfiniteFormat>>) () -> {
                    try {
                        FilePath worldDirectory =
                                regionPath.resolve(name.toLowerCase());
                        FileUtil.createDirectories(worldDirectory);
                        return Optional.of(new BasicTerrainInfiniteFormat(
                                worldDirectory));
                    } catch (IOException e) {
                        LOGGER.error("Error whilst creating world: {}",
                                e.toString());
                    }
                    return Optional.empty();
                });
        if (!format.isPresent()) {
            throw new IOException("Unable to create world");
        }
        WorldServer world =
                new WorldServer(this, name, seed, server.connection(),
                        server.taskExecutor(),
                        newWorld -> new TerrainInfiniteServer(newWorld, 512,
                                format.get(), server.taskExecutor(),
                                plugins.registry().air()), environmentSupplier);
        world.read(worldsTagStructure.getStructure(name));
        return world;
    }

    @Override
    public synchronized void removeWorld(WorldServer world) {
        worldsTagStructure.setStructure(world.id(), world.write());
    }

    @SuppressWarnings({"ReturnOfNull", "CallToNativeMethodWhileLocked"})
    @Override
    public synchronized boolean deleteWorld(String name) {
        FilePath worldDirectory = regionPath.resolve(name.toLowerCase());
        if (FileUtil.exists(worldDirectory)) {
            try {
                AccessController
                        .doPrivileged((PrivilegedExceptionAction<Void>) () -> {
                            FileUtil.deleteDir(worldDirectory);
                            return null;
                        });
            } catch (PrivilegedActionException e) {
                LOGGER.error("Error whilst deleting world: {}", e.toString());
                return false;
            }
        }
        return true;
    }

    @Override
    public void dispose() throws IOException {
        plugins.dispose();
    }

    protected List<PluginFile> pluginFiles() throws IOException {
        FilePath path = this.path.resolve("plugins");
        List<FilePath> files =
                FileUtil.listRecursive(path, FileUtil::isRegularFile,
                        FileUtil::isNotHidden);
        List<PluginFile> plugins = new ArrayList<>(files.size());
        for (FilePath file : files) {
            plugins.add(new PluginFile(file));
        }
        return plugins;
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/server/format/basic/BasicWorldFormat.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.server.format.basic

import mu.KLogging
import org.tobi29.scapes.chunk.EnvironmentServer
import org.tobi29.scapes.chunk.IDStorage
import org.tobi29.scapes.chunk.WorldServer
import org.tobi29.scapes.chunk.terrain.infinite.TerrainInfiniteServer
import org.tobi29.scapes.engine.utils.io.filesystem.*
import org.tobi29.scapes.engine.utils.io.tag.*
import org.tobi29.scapes.engine.utils.task.TaskExecutor
import org.tobi29.scapes.plugins.PluginFile
import org.tobi29.scapes.plugins.Plugins
import org.tobi29.scapes.server.ScapesServer
import org.tobi29.scapes.server.format.PlayerData
import org.tobi29.scapes.server.format.WorldFormat
import java.io.IOException
import java.security.AccessController
import java.security.PrivilegedAction
import java.security.PrivilegedActionException
import java.util.*

class BasicWorldFormat(private val path: FilePath,
                       map: MutableTagMap) : WorldFormat {
    private val regionPath: FilePath
    private val idStorage: IDStorage
    private val plugins: Plugins
    private val playerData: PlayerData
    private val worldsTagStructure: MutableTagMap
    private val seed: Long

    init {
        val security = System.getSecurityManager()
        security?.checkPermission(
                RuntimePermission("scapes.worldFormat"))
        seed = map["Seed"]?.toLong() ?: 0
        idStorage = IDStorage(map.mapMut("IDs"))
        worldsTagStructure = map.mapMut("Worlds")
        regionPath = path.resolve("region")
        playerData = BasicPlayerData(path.resolve("players"))
        plugins = createPlugins()
    }

    private fun createPlugins(): Plugins {
        return Plugins(pluginFiles(), idStorage)
    }

    override fun idStorage(): IDStorage {
        return idStorage
    }

    override fun playerData(): PlayerData {
        return playerData
    }

    override fun seed(): Long {
        return seed
    }

    override fun plugins(): Plugins {
        return plugins
    }

    @Synchronized override fun registerWorld(server: ScapesServer,
                                             environmentSupplier: Function1<WorldServer, EnvironmentServer>,
                                             name: String,
                                             seed: Long): WorldServer {
        val format = AccessController.doPrivileged(
                PrivilegedAction {
                    try {
                        val worldDirectory = regionPath.resolve(
                                name.toLowerCase())
                        createDirectories(worldDirectory)
                        return@PrivilegedAction BasicTerrainInfiniteFormat(
                                worldDirectory)
                    } catch (e: IOException) {
                        logger.error { "Error whilst creating world: $e" }
                    }
                    null
                }) ?: throw IOException("Unable to create world")
        val world = WorldServer(this, name, seed, server.connection,
                TaskExecutor(server.taskExecutor(), name),
                {
                    TerrainInfiniteServer(it, 512, format,
                            it.environment.generator(),
                            arrayOf(it.environment.populator()), it.air)
                }, environmentSupplier)
        worldsTagStructure[name]?.toMap()?.let { world.read(it) }
        return world
    }

    @Synchronized override fun removeWorld(world: WorldServer) {
        worldsTagStructure[world.id] = TagMap { world.write(this) }
    }

    @Synchronized override fun deleteWorld(name: String): Boolean {
        val worldDirectory = regionPath.resolve(name.toLowerCase())
        if (exists(worldDirectory)) {
            try {
                AccessController.doPrivileged(PrivilegedAction {
                    deleteDir(worldDirectory)
                })
            } catch (e: PrivilegedActionException) {
                logger.error { "Error whilst deleting world: $e" }
                return false
            }

        }
        return true
    }

    override fun dispose() {
        plugins.dispose()
    }

    private fun pluginFiles(): List<PluginFile> {
        val path = this.path.resolve("plugins")
        val files = listRecursive(path,
                { isRegularFile(it) && isNotHidden(it) })
        val plugins = ArrayList<PluginFile>(files.size)
        for (file in files) {
            plugins.add(PluginFile(file))
        }
        return plugins
    }

    companion object : KLogging()
}
